### PR TITLE
Upgrade rcf to 4.0

### PR DESCRIPTION
**Pull request by kaituo**
### Description
This PR upgrades rcf to 4.0 as it has bug fixes and support for streaming imputation mode.

Also, 14 out 24 CI items are failing. This PR makes various efforts to fix them.
* fixed tests in ADTaskManagerTests since https://github.com/opensearch-project/anomaly-detection/pull/1131. The failure is likely due to cores changes for the tracing feature.
* fixed flaky test AnomalyDetectorJobTransportActionTests.testStartHistoricalAnalysisForSingleCategoryHCWithUser by changing waiting condition and increase timeout value.
* followed https://github.com/opensearch-project/security/pull/3134/files to resolve bwc CI issue
* merged https://github.com/opensearch-project/anomaly-detection/pull/1163 to fix security CI issue
* followed https://github.com/opensearch-project/anomaly-detection/pull/775 to bump bwc version to latest 2.x
* followed https://github.com/opensearch-project/flow-framework/pull/175/files
  *  error message: '> Task :spotlessJava FAILED
If you upgrade your JVM to 17+, then you can use eclipse jdt formatter 4.29, which may have fixed this problem.
java.lang.Exception: You are running Spotless on JVM 11, which limits you to eclipse jdt formatter 4.26.
If you upgrade your JVM to 17+, then you can use eclipse jdt formatter 4.29, which may have fixed this problem.
28 actionable tasks: 19 executed, 9 up-to-date"
  * cause: This is a bug in spotless: https://github.com/diffplug/spotless/issues/1724
https://github.com/opensearch-project/anomaly-detection/pull/1129 only partially fixed the issue. PR #1129 didn't remove the spotless check from the subsequent build (allowing it to run on JDK11).
* checkout failed due to missing library
  * error message: Run actions/checkout@v4
/usr/bin/docker exec  cdc91410f06d7ab184b358bd420df75fe15958c0071a4fb949ff7724f0ecdcbf sh -c "cat /etc/*release | grep ^ID"
/__e/node20/bin/node: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libc.so.6: version `GLIBC_2.25' not found (required by /__e/node20/bin/node)
  * fix: reverted: https://github.com/opensearch-project/anomaly-detection/pull/1098/files
* followed https://github.com/opensearch-project/OpenSearch/pull/6792/files
  * error message: bwc test would fail with:
    * Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0
  *  fix: replace org.opensearch.gradle.testclusters.StandaloneRestIntegTestTask with  [org.opensearch.gradle.test.RestIntegTestTask](http://org.opensearch.gradle.test.restintegtesttask/) as a temporary workaround for https://docs.gradle.org/8.1/userguide/upgrading_version_8.html#test_task_default_classpath.   
  

Testing done:
1. gradle build

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by joshpalis — APPROVED** [review]
Thanks for fixing the CI @kaituo

File: .github/workflows/test_bwc.yml [hunk 1]
```diff
@@ -38,7 +38,7 @@ jobs:
 
       # anomaly-detection
       - name: Checkout AD
-        uses: actions/checkout@v4
+        uses: actions/checkout@v3
```
Comment (amitgalitz) [inline .github/workflows/test_bwc.yml:41]: why downgrade to checkoutv3?

Comment (kaituo) [reply .github/workflows/test_bwc.yml:41]: mentioned the reason in the Description:

* checkout failed due to missing library
  * error message: Run actions/checkout@v4
/usr/bin/docker exec  cdc91410f06d7ab184b358bd420df75fe15958c0071a4fb949ff7724f0ecdcbf sh -c "cat /etc/*release | grep ^ID"
/__e/node20/bin/node: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libc.so.6: version `GLIBC_2.25' not found (required by /__e/node20/bin/node)
  * fix: reverted: https://github.com/opensearch-project/anomaly-detection/pull/1098/files

Comment (amitgalitz) [reply .github/workflows/test_bwc.yml:41]: sorry missed this

File: src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java [hunk 2]
```diff
@@ -1067,27 +1067,22 @@ public void testDeserializeTRCFModel() throws Exception {
...
-        scores.add(5.770278090352401);
-        scores.add(5.319779117320102);
-
-        List<Double> grade = new ArrayList<>();
-        grade.add(1.0);
-        grade.add(0.0);
-        grade.add(0.0);
-        grade.add(0.0);
-        grade.add(0.0);
+        scores.add(5.052069275347555);
```
Comment (amitgalitz) [inline src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java:1074]: Do you think we should keep the old scores to do have a record of the comparison as we upgrade RCF or that is not really relevant cause point is to test existing functionality, old scores can be seen in commit history basically?

File: build.gradle [hunk 3]
```diff
@@ -149,6 +149,9 @@ dependencies {
         exclude group: 'org.ow2.asm', module: 'asm-tree'
     }
 
+    // used for output encoding of config descriptions
+    implementation group: 'org.owasp.encoder' , name: 'encoder', version: '1.2.3'
```
Comment (amitgalitz) [inline build.gradle:153]: when is this used? I might have missed it but doesn't come up in a quick search

File: src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java (hunk 2 above)
Comment (kaituo) [reply src/test/java/org/opensearch/ad/ml/CheckpointDaoTests.java:1074]: I think the latter. We can find the old scores from commit history.

File: build.gradle (hunk 3 above)
Comment (kaituo) [reply build.gradle:153]: it will be used in later PRs.

**Review by amitgalitz — APPROVED** [review]
LGTM thanks for upgrading!

**Pull request merged**